lib/src/authentication/client.dart

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// ignore_for_file: import_of_legacy_library_into_null_safe
+
+import 'dart:io';
+
+import 'package:http/http.dart' as http;
+
+import '../http.dart';
+import '../system_cache.dart';
+import 'credential.dart';
+import 'credential_store.dart';
+
+/// This client automatically modifies request to contain required credentials
+/// in request. For example some credentials might add `Authentication` header
+/// to request.
+class _AuthenticatedClient extends http.BaseClient {
+  _AuthenticatedClient(
+    this._inner, {
+    required this.credential,
+    required this.serverBaseUrl,
+  });
+
+  final http.BaseClient _inner;
+  final Credential credential;
+  final String serverBaseUrl;
```

Comment:
> Question: do we normalize serverBaseUrl to end in slash or not?

Yes, we do! I'm also normalizing it before matching it with `request.url` to make sure it's normalized if "tokens.json" was modified manually.

Reply:
Okay, hmm, maybe we should do this normalization in:
 * (A) `HostedSource._parseDescription`, and,
 * (B) `HostedSource._pubHostedUrlConfig`, too?

Indeed, correct me if I'm wrong... if we normalize in (A) and (B), then we won't need to normalize anywhere else, right?

Reading (B) it seems like we're already normalizing to **not** have trailing slash. Unless, we have a good reason to change that I propose that we normalize to not have a trailing slash. If we change this will cause unnecessary changes in all `pubspec.lock` files, and they'll swap formatting when switching between Dart SDK versions.

This is perhaps something we should do in a separate PR.

Investigating this I also discovered a bug -> https://github.com/dart-lang/pub/issues/3024.

Reply:
> Reading (B) it seems like we're already normalizing to not have trailing slash.

Actually I don't agree with that.

Wouldn't this might make invalid url matching possible? If pub already authenticated to "http://pub.myserver.com". 
"http://pub.myserver.com.attacker.com" could get token for previously mentioned server because we're matching URLs using `url.startsWith(normalizedBaseUrl)`. Having trailing slash ensures that hostname could not be modified. I don't know how to explain this actually.

You can check this demo regarding my comment: https://dartpad.dev/342b9230acbc6f3527fe4acdd960bab4?null_safety=true